### ISSUE TITLE
support expressions

### ISSUE THREAD

**Issue by rcoh** [rcoh/angle-grinder#13]
eg. `sum(response_ms*2)`. Expressions can be used either inline or as an assignment:
`let response_ms = response_ms * 2`

Probably best to implement in line expressions as a rewrite. Expressions will allow supporting `where`

**Comment by rcoh** [comment]
equality expressions are supported. Arithmetic needed

**Comment by tstack** [comment]
So, I'm taking a whack at this and running into an issue with `eval_borrowed()`: https://github.com/rcoh/angle-grinder/blob/7b356efa1c8beb653808ad85a129c0fbff7f6d49/src/operator.rs#L57

Since the lifetime of the result is the same as the input HashMap, I don't think there's a way to construct an intermediate value for the result of say an addition.  Things have worked so far with `bool` since the true/false values are `&'static`.  I thought about storing the intermediate in the input HashMap, but that is not mutable at this point.  What are your thoughts @rcoh ?  Should the result be cloned instead of returned by ref?  Or, should the input HashMap be made mutable?  Or...

**Comment by rcoh** [comment]
After a brief skim (so this could be totally wrong and I don't remember
exactly why EvalBorrowed exists), can you just implement `Evaluatable`
instead of `EvalutatbleBorrowed`?

**Comment by tstack** [comment]
> can you just implement `Evaluatable` instead of `EvalutatbleBorrowed`?

Hmm, I don't think so... most things seem to call `eval_borrowed()` and it has a big match for the `Expr` enum that needs to have all cases satisfied.  I don't know why `EvaluatableBorrowed` exists either.  Probably for the sake of performance?

**Comment by rcoh** [comment]
Yeah I was probably trying to avoid cloning. It could really mess things up but we could:
- Switch to cloning during eval (RIP eval borrowed)
- Switch to Rc/Arc for the output of eval to avoid a full clone
- Implement Copy for Data::Value either by cloning the string or by sticking the string in an Rc.

**Comment by rcoh** [comment]
OK, I played around with it a little bit. It's a midsize refactor, but I think the answer is to replace `eval_borrowed` with `eval_cow`:
```
pub trait EvaluatableCow<T: Clone> {
    fn eval_cow<'a>(&self, record: &'a Data) -> Result<Cow<'a, T>, EvalError>;
}
```
Cow will allow eval to return something either borrowed or owned and it's pretty ergonomic to use since it implements the std::borrow magic.

I think it may be a pretty big refactoring, so I can take it on if you want.

**Comment by tstack** [comment]
I removed borrowed and have it working.  I'll try to implement the `Cow` version.

**Comment by rcoh** [comment]
yeah either way -- if it doesn't make much of a difference in the
benchmarks just removing borrowed is totally fine.

**Comment by tstack** [comment]
Do you happen to remember what is left to do here?
